File: src/controller.cpp
#include <iostream>
#include <cstring>
#include <WinSock2.h>
#include <WS2tcpip.h>

#pragma comment(lib, "ws2_32.lib")

#define SERVER_IP "192.168.0.X" // Replace with the IP address of the Adafruit Metro M0
#define SERVER_PORT 1234        // Replace with the port number used by the Metro M0 UDP server

int main() {
    WSADATA wsaData;
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        std::cerr << "Failed to initialize Winsock." << std::endl;
        return 1;
    }

    SOCKET sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock == INVALID_SOCKET) {
        std::cerr << "Failed to create socket." << std::endl;
        WSACleanup();
        return 1;
    }

    struct sockaddr_in serverAddress{};
    serverAddress.sin_family = AF_INET;
    serverAddress.sin_port = htons(SERVER_PORT);
    if (inet_pton(AF_INET, SERVER_IP, &(serverAddress.sin_addr)) <= 0) {
        std::cerr << "Failed to convert IP address." << std::endl;
        closesocket(sock);
        WSACleanup();
        return 1;
    }

    char key;
    while (true) {
        std::cout << "Enter W, A, S, or D to control the outputs (Q to quit): ";
        std::cin >> key;

        if (key == 'Q' || key == 'q') {
            break;
        }

        int output;
        switch (key) {
            case 'W':
            case 'w':
                output = 1;
                break;
            case 'A':
            case 'a':
                output = 2;
                break;
            case 'S':
            case 's':
                output = 3;
                break;
            case 'D':
            case 'd':
                output = 4;
                break;
            default:
                std::cout << "Invalid input." << std::endl;
                continue;
        }

        // Send the output number to the UDP server on the Metro M0
        if (sendto(sock, (const char*)&output, sizeof(output), 0, (struct sockaddr*)&serverAddress, sizeof(serverAddress)) == SOCKET_ERROR) {
            std::cerr << "Failed to send data." << std::endl;
            closesocket(sock);
            WSACleanup();
            return 1;
        }
    }

    closesocket(sock);
    WSACleanup();
    return 0;
}

File: src/controller.py
import socket

# Define the target device and port
target_device = "Adafruit Metro M0"
target_port = 6969

# Create a UDP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.settimeout(0.05)  # Set a timeout value to limit the scan duration

# Iterate through IP addresses in the local network
for i in range(1, 255):
    print(i)
    ip_address = f"192.168.0.{i}"  # Customize the IP address range based on your network configuration
    server_address = (ip_address, target_port)

    try:
        # Send a test message to the target device
        sock.sendto("ping".encode(), server_address)
        
        # Wait for a response
        data, addr = sock.recvfrom(1024)
        
        # Check the response to see if it matches the device return message
        if data.decode() == "Pong!":
            print(f"Found {target_device} at IP address: {ip_address}")
            break
            
    except:
        pass

# Close the socket
sock.close()